### PR TITLE
 ⚗️ New decimal-aware price system - Part 4: Add evaluation test for different CT decimals

### PR DESCRIPTION
**Pull request by JuaniRios**
Main info on #260
- Add a test to double check our new business logic on the evaluation round

### CODE REVIEW

**Comment by JuaniRios** [comment]
> [!WARNING]
> <b>This pull request is not mergeable via GitHub because a downstack PR is open. Once all requirements are satisfied, merge this PR as a stack <a href="https://app.graphite.dev/github/pr/Polimec/polimec-node/264?utm_source=stack-comment-downstack-mergeability-warning" >on Graphite</a>.</b>
> <a href="https://graphite.dev/docs/merge-pull-requests">Learn more</a>

* **#264** <a href="https://app.graphite.dev/github/pr/Polimec/polimec-node/264?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* **#263** <a href="https://app.graphite.dev/github/pr/Polimec/polimec-node/263?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#262** <a href="https://app.graphite.dev/github/pr/Polimec/polimec-node/262?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#260** <a href="https://app.graphite.dev/github/pr/Polimec/polimec-node/260?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* `main`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @JuaniRios and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

File: pallets/funding/src/tests/2_evaluation.rs [hunk 1]
```diff
@@ -76,6 +76,158 @@ mod round_flow {
 			let project_status = inst.get_project_details(project_id).status;
 			assert_eq!(project_status, ProjectStatus::FundingFailed);
 		}
+
+		#[test]
+		fn different_decimals_ct_still_works_as_expected() {
```
Comment (vstam1) [inline pallets/funding/src/tests/2_evaluation.rs:81]: Nit: Maybe not call it ct_still_works but rather ct_works. As we know that we have a new decimal system, but after this stack its just the system, so new people coming in might be confused. Like I said small nit.

File: pallets/funding/src/tests/2_evaluation.rs [hunk 2]
```diff
@@ -76,6 +76,158 @@ mod round_flow {
...
+
+			let evaluation_threshold = inst.execute(|| <TestRuntime as Config>::EvaluationSuccessThreshold::get());
+
+			let min_evaluation_amount_usd = <TestRuntime as Config>::MinUsdPerEvaluation::get();
+			let min_evaluation_amount_plmc =
+				usable_plmc_price.reciprocal().unwrap().checked_mul_int(min_evaluation_amount_usd).unwrap();
+
+			// Check the evaluation threshold is calculated correctly for 2 decimals less than USD
+			let mut project_metadata_4_decimals = default_project_metadata.clone();
+			project_metadata_4_decimals.token_information.decimals = 4;
```
Comment (vstam1) [inline pallets/funding/src/tests/2_evaluation.rs:115]: Might be better to use extreme scenario, so decimals = 1? Or even 0 (not sure if allowed)?

File: pallets/funding/src/tests/2_evaluation.rs [hunk 3]
```diff
@@ -76,6 +76,158 @@ mod round_flow {
 			let project_status = inst.get_project_details(project_id).status;
 			assert_eq!(project_status, ProjectStatus::FundingFailed);
 		}
+
+		#[test]
```
Comment (vstam1) [inline pallets/funding/src/tests/2_evaluation.rs:80]: In such a long test can we better define the different stages of the test for clarity? Or if possible split the test up in multiple tests. Either one is good for me.

File: pallets/funding/src/tests/2_evaluation.rs [hunk 4]
```diff
@@ -76,6 +76,158 @@ mod round_flow {
...
+					EVALUATOR_1,
+					InvestorType::Retail,
+					generate_did_from_account(EVALUATOR_1),
+					project_metadata_4_decimals.clone().policy_ipfs_cid.unwrap()
+				),
+				project_id_4_decimals,
+				min_evaluation_amount_usd
+			)));
+
+			// Check that once we bonded the minimum amount, evaluation succeeds
```
Comment (vstam1) [inline pallets/funding/src/tests/2_evaluation.rs:150]: Is it max evaluation that should succeed? So in this case max - min?

**Review by vstam1 — COMMENTED** [review]
Large ass test haha. Left few comments but looks good in general. Might be nice to abstract some pieces away as the 4 and 18 decimal case are 90% the same.

File: pallets/funding/src/tests/2_evaluation.rs (hunk 2 above)
Comment (JuaniRios) [reply pallets/funding/src/tests/2_evaluation.rs:115]: I'll add tests for both 0 and 1 decimal to be safe

File: pallets/funding/src/tests/2_evaluation.rs [hunk 5]
```diff
@@ -76,6 +78,134 @@ mod round_flow {
...
+
+				// The evaluation should succeed when we bond the threshold PLMC amount in total.
+				inst.start_auction(project_id, issuer).unwrap();
+			};
+
+			for decimals in 0..25 {
+				decimal_test(decimals);
+			}
+
+			dbg!(&evaluation_ct_thresholds);
```
Comment (lrazovic) [inline pallets/funding/src/tests/2_evaluation.rs:195]: nit: can we remove the debug statement?

File: pallets/funding/src/tests/2_evaluation.rs [hunk 6]
```diff
@@ -7,6 +7,8 @@ mod round_flow {
 	#[cfg(test)]
 	mod success {
 		use super::*;
+		use futures::StreamExt;
```
Comment (vstam1) [inline pallets/funding/src/tests/2_evaluation.rs:10]: What do we need this import for?

**Review by vstam1 — APPROVED** [review]
Single comment + comment of Leo then ✅

**Review by lrazovic — APPROVED** [review]

**Comment by JuaniRios** [comment]
### Merge activity

* **Apr 30, 9:52 AM EDT**: @JuaniRios started a stack merge that includes this pull request via [Graphite](https://app.graphite.dev/github/pr/Polimec/polimec-node/264).

**Pull request merged**